File: jpa-backend/src/test/java/chaincue/tech/jpabackend/masters/course_master/CourseHelperTest.java
package chaincue.tech.jpabackend.masters.course_master;

import chaincue.tech.jpabackend.masters.student_master.StudentHelper;
import chaincue.tech.jpabackend.masters.teacher_master.TeacherHelper;
import chaincue.tech.jpabackend.utilities.ANSIColors;
import org.junit.jupiter.api.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
class CourseHelperTest {
    @Autowired
    CourseHelper courseHelper;
    @Autowired
    TeacherHelper teacherHelper;
    @Autowired
    StudentHelper studentHelper;

    @Test
    void saveCourseSuccess() {
        var name = courseHelper.saveCourse("name");
        ANSIColors.printBlue(name.getName());
    }

    @Test
    void addTeacherToCourse() {
        //given
        var course = courseHelper.saveCourse("courseName");
        var teacher = teacherHelper.saveTeacher("teacherName");

        //when
        var result = courseHelper.addTeacherToCourse(course.getId(), teacher.getId());
        ANSIColors.printBlue(result);
    }

    @Test
    void addStudentToCourse() {
        //given
        var course = courseHelper.saveCourse("courseName");
        var student = studentHelper.saveStudent("studentName");

        //when
        Course result= courseHelper.addStudentToCourse(course.getId(), student.getId());
        ANSIColors.printBlue(result);
    }

    @Test
    @Transactional
    void getCourseWithTeachers() {
        //given
        var course = courseHelper.saveCourse("courseName");
        var teacher = teacherHelper.saveTeacher("studentName");

        //when
        Course courseWithStudent = courseHelper.addTeacherToCourse(course.getId(), teacher.getId());

        //then
        var result = courseHelper.getCourseById(courseWithStudent.getId()).get();
        ANSIColors.printPurple(result.getTeacherCourseRelations());
    }

    @Test
    @Transactional
    void getCourseWithStudent() {
        //given
        var course = courseHelper.saveCourse("courseName");
        var student1 = studentHelper.saveStudent("studentName");
        var student2 = studentHelper.saveStudent("studentName");

        //when
        courseHelper.addStudentToCourse(course.getId(), student1.getId());
        courseHelper.addStudentToCourse(course.getId(), student2.getId());

        //then
        var result = courseHelper.getCourseById(course.getId()).get();
        ANSIColors.printPurple(result.getStudentCourseRelations());
    }

}

File: kotlin/jpa-backend/src/test/kotlin/chaincue/tech/jpabackend/masters/course_master/CourseHelperTest.kt
package chaincue.tech.jpabackend.masters.course_master

import chaincue.tech.jpabackend.masters.student_master.CourseHelper
import chaincue.tech.jpabackend.masters.student_master.StudentHelper
import chaincue.tech.jpabackend.masters.teacher_master.TeacherHelper
import chaincue.tech.jpabackend.utilities.ANSIColors
import org.junit.jupiter.api.Test

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.annotation.Commit
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
class CourseHelperTest(
        @Autowired val courseHelper: CourseHelper,
        @Autowired val teacherHelper: TeacherHelper,
        @Autowired val studentHelper: StudentHelper,
) {

    @Test
    fun save() {
        val result = courseHelper.save("name")
        ANSIColors.printBlue(result)
    }

    @Test
    @Transactional
    @Commit
    fun findStudent() {
        val course = courseHelper.save("name")
        val result = courseHelper.find(course.id)
        ANSIColors.printBlue(result.get())
    }

    @Test
    @Transactional
    @Commit
    fun addTeacherToCourse() {
        val course = courseHelper.save("names")
        val teacher = teacherHelper.save("name")
        val result = courseHelper.addTeacherToCourse(course.id, teacher.id)
        ANSIColors.printBlue(result)
    }

    @Test
    @Transactional
    @Commit
    fun addStudentToCourse() {
        val course = courseHelper.save("name")
        val student = studentHelper.save("name")
        val result = courseHelper.addStudentToCourse(course.id, student.id)
        ANSIColors.printBlue(result)
    }
}
